File: android-sdk/src/main/java/com/gist/ads/sdk/ui/GistAdControl.kt
package com.gist.ads.sdk.ui

// Main ad control for displaying Gist AI Search ads: embeds the real
// `adtag.js` script in a WebView and drives it via
// `defineSlot({id, api_key, geo}, slotId, sizes, adTypes)` ->
// `slot.definePrompt(query)` -> `displayAd(slotId)`, mirroring exactly how a
// publisher's own webpage would embed the tag directly.
//
// This control is a thin wrapper around that HTML/JS embed: it makes no API
// calls of its own. `adtag.js` owns the entire search request, response
// parsing, and rendering (via its own JSONP GET to the Search API), so as
// the tag and backend evolve, this control keeps working without needing to
// track along. This mirrors the pure-embed pattern already shipped for
// `GistDisplayAdControl` -- see that file's header comment and the plan
// that migrated search ads to match it.
//
// Because `adtag.js`'s own search request has no concept of a selectable
// API version, the previous `apiVersion` parameter has been dropped
// entirely (no v1 fallback), matching the precedent set when `context` was
// dropped from display ads. `sizes` is a new public param since
// `defineSlot` requires a non-empty `sizes` array and the previous
// iframe-based control had no sizing concept at all.
//
// Note on the publisher-key exposure trade-off: `publisherKey` is now
// embedded directly in the bootstrap HTML/JS and sent by `adtag.js` as a
// `publisher_key` query param in a public JSONP GET made from inside the
// WebView -- visible in the loaded HTML/JS source (inspectable via WebView
// dev tools). This is the same exposure model a publisher already accepts
// by embedding the JS tag on a public webpage; native apps lose the extra
// native-only protection they previously had from sending it as a hidden
// HTTP header in a server-side POST.

import androidx.compose.foundation.isSystemInDarkTheme
import androidx.compose.foundation.layout.*
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Warning
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import com.gist.ads.sdk.APIConstants
import com.gist.ads.sdk.models.AdSize
import com.gist.ads.sdk.models.AdTagLoadState
import com.gist.ads.sdk.models.AdType
import com.gist.ads.sdk.utils.SearchAdBootstrapHTML
import java.util.UUID
import kotlinx.coroutines.launch

/**
 * Main composable control for displaying Gist AI Search ads.
 *
 * Unlike [GistDisplayAdControl] (display ads, no secret key), search ads are
 * gated by a secret publisher key and targeted by a search query. Mirrors
 * the web tag's `defineSlot({id, api_key, geo}, slotId, sizes, adTypes)` ->
 * `slot.definePrompt(query)` -> `displayAd(slotId)` flow.
 *
 * @param publisherId Your publisher ID credential
 * @param publisherKey Your publisher API key
 * @param query Search query to fetch relevant ads for (mirrors `slot.definePrompt(...)`)
 * @param geo Geographic location code (e.g., "US", "GB", "CA")
 * @param answer Answer text (mirrors `slot.defineAnswer(...)`); the ad tag docs mark this as
 *   required for search ads, so when omitted it defaults to [query]
 * @param adTypes Optional list of ad types to filter (null = all types)
 * @param sizes One or more supported ad sizes (mirrors `sizes` in `defineSlot`); defaults to
 *   `[AdSize.DYNAMIC]`
 * @param environment API environment (defaults to production)
 * @param modifier Modifier for styling the ad control
 * @param theme Theme preference - "light", "dark", or "system" (defaults to "system" for auto-detection)
 * @param onAdLoaded Optional callback invoked when ad successfully loads
 * @param onAdClicked Optional callback invoked when user clicks an ad link (provides URL); if
 *   null, opens the URL in the default browser
 * @param onContentHeightChanged Optional callback invoked when ad content height is measured
 * @param passback Composable shown when no ad is available (no-fill), mirroring the web tag's
 *   `definePassbackFunction`: callers get full control over the fallback UI instead of a
 *   hard-coded empty state. Defaults to a simple "No ad available" text when not provided.
 */
@Composable
fun GistAdControl(
    publisherId: String,
    publisherKey: String,
    query: String,
    geo: String = "US",
    answer: String? = null,
    adTypes: List<AdType>? = null,
    sizes: List<AdSize> = listOf(AdSize.DYNAMIC),
    environment: APIConstants.Environment = APIConstants.Environment.PRODUCTION,
    modifier: Modifier = Modifier,
    theme: String = "system",
    onAdLoaded: (() -> Unit)? = null,
    onAdClicked: ((String) -> Unit)? = null,
    onContentHeightChanged: ((Float) -> Unit)? = null,
    passback: (@Composable () -> Unit)? = null
) {
    val isDarkMode = isSystemInDarkTheme()

    // Resolve theme to "light" or "dark".
    //
    // WORKAROUND: PrtAdsTag's CSS for `data-theme="dark"` sets
    // `color-scheme: dark` on `:root` while the wrapper is transparent and
    // `.content-container` keeps explicit `color: #000000` for text. This
    // produces black text on a dark canvas (invisible) for any non-`gist.ai`
    // publisher. Until PrtAdsTag adds proper dark theme support for
    // arbitrary publishers, we coerce "dark" to "light" here so the ad
    // renders correctly. The surrounding app UI is unaffected — it still
    // follows `isSystemInDarkTheme()` / Material theming.
    val resolvedTheme = remember(theme, isDarkMode) {
        val requested = when (theme) {
            "light" -> "light"
            "dark" -> "dark"
            "system" -> if (isDarkMode) "dark" else "light"
            else -> if (isDarkMode) "dark" else "light"
        }
        if (requested == "dark") "light" else requested
    }

    var state by remember { mutableStateOf<AdTagLoadState>(AdTagLoadState.Loading) }
    var currentSlot by remember { mutableStateOf<SearchAdSlotLoad?>(null) }

    val scope = rememberCoroutineScope()

    // Identity key capturing every load-triggering parameter, mirroring
    // GistDisplayAdControl's `loadKey`: `LaunchedEffect` below re-runs
    // `prepareSlot()` whenever this changes.
    val loadKey = remember(query, geo, answer, adTypes, sizes, environment, resolvedTheme) {
        listOf(
            query,
            geo,
            answer ?: "",
            adTypes?.joinToString(",") { it.value } ?: "",
            sizes.joinToString(",") { it.displayName },
            environment.name,
            resolvedTheme
        ).joinToString("|")
    }

    // Prepare (or re-prepare, on retry/param change) the WebView's bootstrap
    // HTML: mint a fresh slot id + passback function name and build the
    // bootstrap HTML. This is purely local string generation -- no network
    // call is made here or anywhere else in this control; `adtag.js` makes
    // its own request once the resulting WebView loads.
    fun prepareSlot() {
        val slotId = "pr-search-ad-${UUID.randomUUID()}"
        val passbackFunctionName = "prSearchAdPassback${UUID.randomUUID().toString().replace("-", "")}"
        state = AdTagLoadState.Loading
        currentSlot = null

        try {
            val html = SearchAdBootstrapHTML.generate(
                publisherId = publisherId,
                publisherKey = publisherKey,
                query = query,
                geo = geo,
                answer = answer,
                adTypes = adTypes,
                sizes = sizes,
                slotId = slotId,
                passbackFunctionName = passbackFunctionName,
                adTagScriptUrl = environment.adTagScriptUrl
            )
            currentSlot = SearchAdSlotLoad(slotId, html)
        } catch (e: Exception) {
            state = AdTagLoadState.Failed(e.message ?: "Unknown error")
        }
    }

    LaunchedEffect(loadKey) {
        prepareSlot()
    }

    Box(
        modifier = modifier.fillMaxWidth(),
        contentAlignment = Alignment.Center
    ) {
        when (val currentState = state) {
            is AdTagLoadState.Loading, is AdTagLoadState.Loaded -> {
                currentSlot?.let { slot ->
                    key(slot.slotId) {
                        AdTagBridgeWebView(
                            html = slot.html,
                            baseUrl = environment.adTagScriptUrl,
                            modifier = Modifier.fillMaxWidth(),
                            theme = resolvedTheme,
                            onAdRendered = { height ->
                                state = AdTagLoadState.Loaded(height)
                                onContentHeightChanged?.invoke(height.toFloat())
                                onAdLoaded?.invoke()
                            },
                            onNoFill = {
                                state = AdTagLoadState.NoFill
                            },
                            onLoadFailure = { message ->
                                state = AdTagLoadState.Failed(message)
                            },
                            onAdClicked = onAdClicked
                        )
                    }
                }
                if (currentState is AdTagLoadState.Loading) {
                    SearchAdLoadingView()
                }
            }
            is AdTagLoadState.Failed -> {
                SearchAdErrorView(
                    message = currentState.message,
                    onRetry = { scope.launch { prepareSlot() } }
                )
            }
            is AdTagLoadState.NoFill -> {
                if (passback != null) {
                    passback()
                } else {
                    SearchAdDefaultNoFillView()
                }
            }
        }
    }
}

private data class SearchAdSlotLoad(val slotId: String, val html: String)

@Composable
private fun SearchAdLoadingView() {
    Column(
        modifier = Modifier
            .fillMaxWidth()
            .padding(32.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.Center
    ) {
        CircularProgressIndicator()
        Spacer(modifier = Modifier.height(16.dp))
        Text(
            text = "Loading ad...",
            style = MaterialTheme.typography.bodyMedium
        )
    }
}

@Composable
private fun SearchAdErrorView(
    message: String,
    onRetry: () -> Unit
) {
    Column(
        modifier = Modifier
            .fillMaxWidth()
            .padding(24.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.Center
    ) {
        Icon(
            imageVector = Icons.Default.Warning,
            contentDescription = "Error",
            modifier = Modifier.size(48.dp),
            tint = MaterialTheme.colorScheme.error
        )

        Spacer(modifier = Modifier.height(16.dp))

        Text(
            text = "Unable to load ad",
            style = MaterialTheme.typography.titleMedium,
            color = MaterialTheme.colorScheme.onSurface
        )

        Spacer(modifier = Modifier.height(8.dp))

        Text(
            text = message,
            style = MaterialTheme.typography.bodySmall,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
            textAlign = androidx.compose.ui.text.style.TextAlign.Center
        )

        Spacer(modifier = Modifier.height(16.dp))

        Button(onClick = onRetry) {
            Text("Retry")
        }
    }
}

@Composable
private fun SearchAdDefaultNoFillView() {
    Box(
        modifier = Modifier
            .fillMaxWidth()
            .padding(32.dp),
        contentAlignment = Alignment.Center
    ) {
        Text(
            text = "No ad available",
            style = MaterialTheme.typography.bodyMedium,
            color = MaterialTheme.colorScheme.onSurfaceVariant
        )
    }
}

File: swift-sdk/Sources/GistAdsSDK/Views/GistAdControl.swift
//
//  GistAdControl.swift
//  GistAdsSDK
//
//  Main ad control for displaying Gist AI Search ads: embeds the real
//  `adtag.js` script in a WebView and drives it via
//  `defineSlot({id, api_key, geo}, slotId, sizes, adTypes)` ->
//  `slot.definePrompt(query)` -> `displayAd(slotId)`, mirroring exactly how
//  a publisher's own webpage would embed the tag directly for search.
//
//  This control is a thin wrapper around that HTML/JS embed: it makes no
//  API calls of its own. `adtag.js` owns the entire ad request (its own
//  JSONP GET to the Search API), response parsing, and rendering directly
//  into the slot's DOM (no iframe), so as the tag and backend evolve, this
//  control keeps working without needing to track along. (An earlier
//  revision fetched ads natively via `AdAPIService` and wrapped the
//  response's `iframeUrl` in an `<iframe>`; that path has been removed so
//  this control can be a pure embed, mirroring `GistDisplayAdControl` --
//  see PR #7 review discussion and the follow-up search-ad migration.)
//
//  Note: unlike display ads, search ads are gated by a secret
//  `publisherKey`. Embedding it here means it becomes visible in the
//  loaded HTML/JS source and is sent as a public `publisher_key` query
//  param in the tag's own JSONP request -- the same exposure a publisher
//  already accepts by embedding the JS tag on a public webpage. Native
//  apps lose the extra protection of keeping it server-side/header-only.
//

import SwiftUI

/// Main control for displaying Gist AI Search ads.
public struct GistAdControl: View {

    /// Environment configuration for API endpoints
    public enum APIEnvironment {
        case staging
        case integration
        case production

        /// Iframe base URL for the environment (internal). Also used as the
        /// base URL for the embedded `adtag.js` bundle.
        /// Can be overridden via environment variables: GIST_ADS_STAGING_IFRAME_URL, GIST_ADS_INTEGRATION_IFRAME_URL, GIST_ADS_PRODUCTION_IFRAME_URL
        internal var iframeBaseURL: String {
            APIConstants.iframeBaseURL(for: self)
        }
    }

    // MARK: - Configuration Properties

    private let publisherID: String
    private let publisherKey: String
    private let query: String
    private let geo: String
    private let answer: String?
    private let adTypes: [AdType]?
    private let sizes: [AdSize]
    private let environment: APIEnvironment
    private let theme: String

    // MARK: - Callbacks

    /// Called when the ad has successfully loaded and is ready to display.
    private let onAdLoaded: (() -> Void)?

    /// Called when the user clicks an ad link that navigates to an external URL.
    /// If not provided, the URL will be opened in the default browser.
    private let onAdClicked: ((URL) -> Void)?

    /// Called when the ad content has loaded and the actual content height is known.
    /// Use this to resize your ad container to match the actual ad content size.
    private let onContentHeightChanged: ((CGFloat) -> Void)?

    /// View builder invoked when no ad is available (no-fill), mirroring the
    /// web tag's `definePassbackFunction`: the caller gets full control over
    /// what to show instead of the ad, rather than a hard-coded empty state.
    private let passback: () -> AnyView

    // MARK: - State

    private struct SearchAdSlotLoad: Equatable {
        let slotID: String
        let html: String
    }

    @State private var state: AdTagLoadState = .loading
    @State private var currentSlot: SearchAdSlotLoad?
    @Environment(\.colorScheme) var colorScheme: ColorScheme

    /// URL of the `adtag.js` bundle to embed for this environment. Reuses
    /// the same host as display ads, since `adtag.js` is one bundle serving
    /// both ad types (distinguished by whether `api_key` is passed to
    /// `defineSlot`). Also used as the WebView's `baseURL` for
    /// `loadHTMLString`, which is why this host must stay in
    /// `APIConstants.allowedAdDomains`.
    private var adTagScriptURL: String {
        let base = APIConstants.iframeBaseURL(for: environment)
        return base.replacingOccurrences(of: "/$", with: "", options: .regularExpression) + "/adtag.js"
    }

    // MARK: - Initialization

    /// Initialize the Gist ad control with custom no-fill content.
    /// - Parameters:
    ///   - publisherID: Your publisher ID
    ///   - publisherKey: Your publisher API key
    ///   - query: The search query to fetch ads for (passed via `slot.definePrompt(...)`)
    ///   - geo: Geographic location code (e.g., "US", "GB")
    ///   - answer: Answer text (passed via `slot.defineAnswer(...)`); the ad tag docs mark this
    ///     as required for search ads, so when omitted it defaults to `query`
    ///   - adTypes: Optional array of ad types to filter (defaults to all types)
    ///   - sizes: One or more supported ad sizes (mirrors `sizes` in `defineSlot`; defaults to `[.dynamic]`)
    ///   - environment: API environment (defaults to production)
    ///   - theme: Theme preference - "light", "dark", or "system" (defaults to "system" for auto-detection)
    ///   - onAdLoaded: Optional callback when ad successfully loads
    ///   - onAdClicked: Optional callback when user clicks an ad link (defaults to opening in browser)
    ///   - onContentHeightChanged: Optional callback when ad content height is determined
    ///   - passback: View builder shown when no ad is available (no-fill), mirroring `definePassbackFunction`
    public init<Passback: View>(
        publisherID: String,
        publisherKey: String,
        query: String,
        geo: String = "US",
        answer: String? = nil,
        adTypes: [AdType]? = nil,
        sizes: [AdSize] = [.dynamic],
        environment: APIEnvironment = .production,
        theme: String = "system",
        onAdLoaded: (() -> Void)? = nil,
        onAdClicked: ((URL) -> Void)? = nil,
        onContentHeightChanged: ((CGFloat) -> Void)? = nil,
        @ViewBuilder passback: @escaping () -> Passback
    ) {
        self.publisherID = publisherID
        self.publisherKey = publisherKey
        self.query = query
        self.geo = geo
        self.answer = answer
        self.adTypes = adTypes
        self.sizes = sizes
        self.environment = environment
        self.theme = theme
        self.onAdLoaded = onAdLoaded
        self.onAdClicked = onAdClicked
        self.onContentHeightChanged = onContentHeightChanged
        self.passback = { AnyView(passback()) }
    }

    /// Initialize the Gist ad control with a built-in "No ad available" fallback.
    /// - Parameters:
    ///   - publisherID: Your publisher ID
    ///   - publisherKey: Your publisher API key
    ///   - query: The search query to fetch ads for
    ///   - geo: Geographic location code (e.g., "US", "GB")
    ///   - answer: Answer text; the ad tag docs mark this as required for search ads, so when
    ///     omitted it defaults to `query`
    ///   - adTypes: Optional array of ad types to filter (defaults to all types)
    ///   - sizes: One or more supported ad sizes (defaults to `[.dynamic]`)
    ///   - environment: API environment (defaults to production)
    ///   - theme: Theme preference - "light", "dark", or "system" (defaults to "system" for auto-detection)
    ///   - onAdLoaded: Optional callback when ad successfully loads
    ///   - onAdClicked: Optional callback when user clicks an ad link (defaults to opening in browser)
    ///   - onContentHeightChanged: Optional callback when ad content height is determined
    public init(
        publisherID: String,
        publisherKey: String,
        query: String,
        geo: String = "US",
        answer: String? = nil,
        adTypes: [AdType]? = nil,
        sizes: [AdSize] = [.dynamic],
        environment: APIEnvironment = .production,
        theme: String = "system",
        onAdLoaded: (() -> Void)? = nil,
        onAdClicked: ((URL) -> Void)? = nil,
        onContentHeightChanged: ((CGFloat) -> Void)? = nil
    ) {
        self.init(
            publisherID: publisherID,
            publisherKey: publisherKey,
            query: query,
            geo: geo,
            answer: answer,
            adTypes: adTypes,
            sizes: sizes,
            environment: environment,
            theme: theme,
            onAdLoaded: onAdLoaded,
            onAdClicked: onAdClicked,
            onContentHeightChanged: onContentHeightChanged,
            passback: { GistAdControl.defaultNoFillView }
        )
    }

    private static var defaultNoFillView: some View {
        Text("No ad available")
            .foregroundColor(.secondary)
            .frame(maxWidth: .infinity, maxHeight: .infinity)
    }

    // MARK: - Body

    public var body: some View {
        ZStack {
            switch state {
            case .loading, .loaded:
                if let currentSlot {
                    AdTagBridgeWebView(
                        html: currentSlot.html,
                        baseURLString: adTagScriptURL,
                        theme: resolvedTheme,
                        onAdRendered: { height in
                            state = .loaded(height: height)
                            onContentHeightChanged?(CGFloat(height))
                            onAdLoaded?()
                        },
                        onNoFill: {
                            state = .noFill
                        },
                        onLoadFailure: { message in
                            state = .failed(message)
                        },
                        onAdClicked: onAdClicked
                    )
                    .id(currentSlot.slotID)
                    .opacity(isLoaded ? 1 : 0)
                }
                if !isLoaded {
                    ProgressView("Loading ad...")
                }
            case .noFill:
                passback()
            case .failed(let message):
                errorView(message: message)
            }
        }
        .frame(maxWidth: .infinity)
        .frame(minHeight: AdViewConstants.configurableMinHeight, maxHeight: AdViewConstants.configurableMaxHeight)
        .task(id: loadKey) {
            prepareSlot()
        }
    }

    private var isLoaded: Bool {
        if case .loaded = state { return true }
        return false
    }

    // MARK: - Views

    private func errorView(message: String) -> some View {
        VStack(spacing: 12) {
            Image(systemName: "exclamationmark.triangle")
                .font(.largeTitle)
                .foregroundColor(.orange)

            Text("Unable to load ad")
                .font(.headline)

            Text(message)
                .font(.caption)
                .foregroundColor(.secondary)
                .multilineTextAlignment(.center)

            Button("Retry") {
                prepareSlot()
            }
            .buttonStyle(.borderedProminent)
        }
        .padding()
        .frame(maxWidth: .infinity, maxHeight: .infinity)
    }

    // MARK: - Methods

    /// Resolve theme to "light" or "dark" based on preference and system settings.
    ///
    /// WORKAROUND: PrtAdsTag's CSS for `data-theme="dark"` sets
    /// `color-scheme: dark` on `:root` while the wrapper is transparent and
    /// `.content-container` keeps explicit `color: #000000` for text. This
    /// produces black text on a dark canvas (invisible) for any non-`gist.ai`
    /// publisher. Until PrtAdsTag adds proper dark theme support for
    /// arbitrary publishers, we coerce "dark" to "light" here so the ad
    /// renders correctly. The surrounding app UI is unaffected -- it can
    /// still be styled dark via SwiftUI's `.preferredColorScheme`.
    private var resolvedTheme: String {
        let requested: String
        switch theme {
        case "light": requested = "light"
        case "dark": requested = "dark"
        case "system": requested = colorScheme == .dark ? "dark" : "light"
        default: requested = colorScheme == .dark ? "dark" : "light"
        }
        return requested == "dark" ? "light" : requested
    }

    /// Identity key for `.task(id:)`, capturing every parameter that should
    /// trigger a reload. SwiftUI recreates this struct with new parameter
    /// values whenever a caller updates them (e.g. `query` for a new search),
    /// but `.task` alone only fires once for the view's lifetime -- it needs
    /// an explicit `id` to react to those changes.
    var loadKey: String {
        [
            query,
            geo,
            answer ?? "",
            (adTypes?.map(\.rawValue).joined(separator: ",")) ?? "",
            sizes.map(\.displayName).joined(separator: ","),
            String(describing: environment),
            resolvedTheme
        ].joined(separator: "|")
    }

    /// Prepare (or re-prepare, on retry/param change) the WebView's bootstrap
    /// HTML: mints a fresh slot id + passback function name and builds the
    /// bootstrap HTML. This is purely local string generation -- no network
    /// call is made here or anywhere else in this control; `adtag.js` makes
    /// its own request once the resulting WebView loads. Mounting that
    /// WebView and waiting for its bridge signals is what drives `state` to
    /// `.loaded`/`.noFill` from there.
    private func prepareSlot() {
        let slotID = "pr-search-ad-\(UUID().uuidString)"
        let passbackFunctionName = "prSearchAdPassback\(UUID().uuidString.replacingOccurrences(of: "-", with: ""))"
        state = .loading
        currentSlot = nil

        do {
            let html = try SearchAdBootstrapHTML.generate(
                publisherID: publisherID,
                publisherKey: publisherKey,
                query: query,
                geo: geo,
                answer: answer,
                adTypes: adTypes,
                sizes: sizes,
                slotID: slotID,
                passbackFunctionName: passbackFunctionName,
                adTagScriptURL: adTagScriptURL
            )
            currentSlot = SearchAdSlotLoad(slotID: slotID, html: html)
        } catch {
            state = .failed(error.localizedDescription)
        }
    }

    /// Reload the ad with current configuration
    public func reload() async {
        prepareSlot()
    }
}

// MARK: - Convenience Initializers

extension GistAdControl {
    /// Initialize with specific ad types
    public static func withAdTypes(
        publisherID: String,
        publisherKey: String,
        query: String,
        geo: String = "US",
        adTypes: [AdType],
        sizes: [AdSize] = [.dynamic],
        environment: APIEnvironment = .production,
        onAdLoaded: (() -> Void)? = nil,
        onAdClicked: ((URL) -> Void)? = nil,
        onContentHeightChanged: ((CGFloat) -> Void)? = nil,
        theme: String = "system"
    ) -> GistAdControl {
        GistAdControl(
            publisherID: publisherID,
            publisherKey: publisherKey,
            query: query,
            geo: geo,
            adTypes: adTypes,
            sizes: sizes,
            environment: environment,
            theme: theme,
            onAdLoaded: onAdLoaded,
            onAdClicked: onAdClicked,
            onContentHeightChanged: onContentHeightChanged
        )
    }
}
